File: KTH_DS02/somdtw.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include "somdtw.h"
#include "dtw.h"

#define VINF 1e15



void init_conf(int channel, int height, int width, int n_iter, int w, int b, double lr, double sigma, t_conf *NetConf)
{
    NetConf->channel = channel;
    NetConf->height = height; // row, x
    NetConf->width = width;  // column, y
    NetConf->neuron = height * width;
    NetConf->n_iter = n_iter;
    NetConf->w = w;
    NetConf->b = b;
    NetConf->lr = lr;
    NetConf->sigma = sigma;
}

/*
void alloc_array_struct(int n)
{
    d_array=malloc(n*sizeof(Data));
    int i;
    for(i=0;i<n;i++)
    {
        d_array[i].data=malloc(N_conf.n_feature*sizeof(double));
        d_array[i].cluster=malloc(sizeof(int));
    }
}
*/

double euc_dist1(double *x, double *y, int xlen)
{
    double sum = 0.0;
    int i;
    for(i = 0; i < xlen; ++i)
    {
        sum += pow((x[i]-y[i]),2.0);
    }
    return sqrt(sum);
}


double* init_rand_weight(double rg, int ch)
{
    int i;
    double k = (double) rand()/RAND_MAX;
    double *tmp_w = malloc(ch*sizeof(double));

    for(i=0; i<ch; i++)
    {
        tmp_w[i] = k * rg + rg/2;
    }

    double norm=0.;

    for(i=0; i<ch; i++)
    {
        norm += pow(tmp_w[i], 2.0);
    }

    for(i=0; i<ch; i++)
    {
            tmp_w[i] /= norm;
    }
    return tmp_w;

}

//array shuffle

void init_shuffle(int *index_array, int n)
{
    int i;
    for(i=0;i<n;i++)
    {
        index_array[i] = i;
    }
    return;
}


void array_shuffle(int *index_array, int n)
{
    int i,ri,k;
    srand(time(NULL));
    for(i=n-1; i>=1; i--)
    {
        ri = (int) rand() % i;
        k = index_array[i];
        index_array[i] = index_array[ri];
        index_array[ri] = k;
    }

    return;
}


// create and initialize map
void create_neuron_map(t_net *Net, t_conf *NetConf)
{

    int i, j;
    Net->map = malloc(NetConf->height * sizeof(t_node *));
	for(i=0; i<NetConf->height; i++)
	{
		Net->map[i] = malloc(NetConf->width * sizeof(t_node));
	}
	for(i=0; i<NetConf->height; i++)
	{
		for (j=0; j<NetConf->width; j++)
		{
            Net->map[i][j].w = (double*)malloc(sizeof(double)*NetConf->channel);
			Net->map[i][j].w = init_rand_weight(0.005, NetConf->channel);
		}
	}
	Net->nb_size = NetConf->sigma;

}

void update(t_bmu *bn, t_net *Net, t_conf *NetConf)
{
    int ns = Net->nb_size;
    int i,j,k,x1,y1,x2,y2; // bubble size
    for(; ns>=0; ns--)
    {
        if(bn->r - ns < 0) x1 = 0;
        else x1 = bn->r - ns;
        if(bn->c - ns < 0) y1 = 0;
        else y1 = bn->c - ns;

        if(bn->r + ns > NetConf->height - 1) x2 = NetConf->height - 1;
        else x2 = bn->r + ns;
        if(bn->c + ns > NetConf->width - 1) y2 = NetConf->width - 1;
        else y2 = bn->c + ns;
        // bubble update
        //printf("i: %d; j: %d; k: %d. \n", x2 ,y2, x1);
        for(i=x1; i<=x2; i++)
        {
            for(j=y1; j<=y2; j++)
            {
                for(k=0; k<NetConf->channel; k++)
                {
                    Net->map[i][j].w[k] += Net->lr * (Net->data[k] - Net->map[i][j].w[k]);
                }
            }
        }
    }
    return;
}


double lr_decay(double olr, int it, int n_it)
{
    double lr;
    lr = olr / (1.0 + (double)it/((double)n_it/2));
    return lr;
}


int nb_decay(double sigma, int it, int n_it)
{
    double p;
    int k;
    p = sigma / (1.0 + (double)it/((double)n_it/2));
    if (p > 1.0)
    {
        k = (int) floor(p);
    }
    else k = 1;

    return k;
}


void train(t_bmu *classification, double *data, int n_samples, int m_features, t_net *Net, t_conf *NetConf, int mode)
{
    double dist, min_dist, total_dist;
    int i, j, it, s, idx;
    int cx, cy;
    t_bmu *BMU = malloc(sizeof(t_bmu));

    int *index_array = malloc(sizeof(int)*n_samples);

    init_shuffle(index_array, n_samples);
    double t1 = 0;
    for(it=0; it<NetConf->n_iter; it++)
    {
        total_dist = 0;
        Net->lr = lr_decay(NetConf->lr, it, NetConf->n_iter);
        Net->nb_size = nb_decay(NetConf->sigma, it, NetConf->n_iter);
        array_shuffle(index_array, n_samples);

        clock_t t_start,t_end;
        t_start = clock();

        for(s=0; s<n_samples; s++)
        {
            idx = index_array[s];
            Net->data = data + idx * m_features;
            min_dist = VINF;
            cx = 0;
            cy = 0;
            // find bmu
            //printf("%d %d abc.\n", NetConf->height, NetConf->width);
            //printf("%d %d def.\n", NetConf->n_iter, s);
            if(mode == 1){
                for(i=0; i < NetConf->height; i++)
                {
                    for(j=0; j < NetConf->width; j++)
                    {
                        //printf("%d %d abc.\n", i, j);
                        dist = LB_Keogh(Net->data, Net->map[i][j].w, m_features, m_features, NetConf->b);
                        if(dist < min_dist)
                        {
                            dist = dtw(Net->data, Net->map[i][j].w, m_features, m_features, NetConf->w);
                            if(dist < min_dist)
                            {
                                min_dist = dist;
                                cx = i;
                                cy = j;
                            }
                        }

                    }
                }
            }
            else if(mode == 2){
                for(i=0; i < NetConf->height; i++)
                {
                    for(j=0; j < NetConf->width; j++)
                    {
                        //printf("%d %d abc.\n", i, j);
                        dist = euc_dist1(Net->data, Net->map[i][j].w, m_features);
                        if(dist < min_dist)
                        {
                            min_dist = dist;
                            cx = i;
                            cy = j;
                        }

                    }
                }
            }
            else
            {
                printf("Wrong mode.\n");
                free(BMU);
                free(index_array);
                return;
            }

            BMU->r = cx;
            BMU->c = cy;
            BMU->dist = min_dist;
            total_dist += min_dist;
            classification[idx].r = cx;
            classification[idx].c = cy;
            classification[idx].dist = min_dist;

        }
        t_end = clock();
        printf("Time=%f, cycle = %d.\n",(double)(t_end-t_start)/CLOCKS_PER_SEC, it+1);
        t1 += (double)(t_end-t_start)/CLOCKS_PER_SEC;
        update(BMU, Net, NetConf);
        total_dist /= n_samples;
        printf("Total distance in %d iteration is %f.\n", it+1, total_dist);


    }
    printf("Time=%f.\n",(double)(t1/NetConf->n_iter));
    free(BMU);
    free(index_array);
    return;
}


void somtrain(t_bmu *classification, double *data, int n_samples, int m_features, int height,
              int width, int n_iter, int w, int b, double lr, double sigma, int mode)
{
     t_conf *NetConf = malloc(sizeof(t_conf));
     t_net *Net = malloc(sizeof(t_net));
     init_conf(m_features, height, width, n_iter, w, b, lr, sigma, NetConf);
     printf("Initialization.\n");
     create_neuron_map(Net, NetConf);
     printf("Create map.\n");
     train(classification, data, n_samples, m_features, Net, NetConf, mode);
     printf("Training process complete.\n");
     free(NetConf);
     free(Net);
     return;

}



/*
int main()
{
    double data[2][128] = {
        {0.037764633,2.5311236E-4,0.003977076,0.020764058,0.02173118,0.114,0.414,0.718,0.889,0.969,1.002,1.014,1.017,0.905,0.605,0.304,0.138,0.061,0.053973303,0.0146603,0.035561278,0.045494101,0.030238212,0.043190555,0.021148818,0.039419722,0.017780126,0.04535716,0.04564139,0.018776462,0.012347944,0.02442172,0.018924466,0.0069161007,0.042104272,5.1942129E-4,-0.114,-0.399,-0.682,-0.836,-0.906,-0.933,-0.943,-0.944,-0.944,-0.948,-0.949,-0.946,-0.948,-0.951,-0.952,-0.952,-0.953,-0.952,-0.952,-0.952,-0.95,-0.949,-0.952,-0.956,-0.96,-0.961,-0.959,-0.958,-0.96,-0.961,-0.96,-0.96,-0.959,-0.958,-0.956,-0.953,-0.949,-0.947,-0.948,-0.951,-0.95,-0.945,-0.941,-0.941,-0.943,-0.944,-0.943,-0.944,-0.952,-0.957,-0.953,-0.946,-0.838,-0.56,-0.284,-0.131,-0.057,-0.015058931,0.0059390887,0.0022057233,0.0083295739,0.028908968,0.0049785927,0.041239881,0.010012772,0.019281287,0.033087255,0.0099204681,0.026583031,0.0094319071,0.018443621,0.023889741,0.043350606,0.030228536,0.031086542,0.012996399,0.034532338,0.03117658,0.019603584,0.027655279,0.041001638,0.019075367,0.034208601,0.022830972,0.013387386,0.038906948,0.039187692,0.053616705,0.054673016,0.031523157,0.03978737,0.020660802},
        {0.041747222,0.026262035,0.0024890674,0.023925895,0.039603507,0.039008461,0.021147631,0.018765444,0.026405114,0.020522961,0.023448213,0.0018729446,0.003766064,0.030746577,0.04354786,0.034775415,0.046251109,0.040955577,0.018543766,0.016918974,0.025676501,0.020744765,0.01273871,0.048883031,0.010178469,0.0087539087,0.020608857,0.026569593,0.038964386,0.0020765946,0.026232962,0.010357591,0.01102224,0.026302193,0.032369582,0.0439846,0.032001582,0.013770521,0.032339293,0.0099025671,0.117,0.423,0.731,0.902,0.979,1.013,1.027,1.03,1.032,1.036,1.035,1.035,1.036,1.037,1.04,1.043,1.043,1.04,1.039,1.041,1.04,1.036,1.035,1.034,1.031,1.031,1.034,1.039,1.041,1.039,1.034,1.029,1.029,1.033,1.035,1.036,1.034,1.034,1.04,0.927,0.618,0.308,0.137,0.057,0.052822401,0.03712816,0.028541798,0.025232772,0.05503598,0.015742396,0.011606101,0.039391864,0.03830194,0.049148719,0.019788488,0.023780675,0.049672354,0.037647625,0.050046987,0.048581181,0.026163674,0.029072628,0.026550945,0.038327926,0.032552285,0.052308944,0.035750869,0.03598284,0.027195561,0.010571572,7.1233322E-4,0.036944592,0.0045633801,0.039512027,0.042119848,0.04212316,0.035241181,0.0057523678,0.028412641,0.040423775,0.028007071,0.042979278,0.0078219738,0.013169194,0.022393536,0.037013136,0.030751863,0.053828632}
    };
    double *p = data;
    init_conf(channel=128, height=3, width=3, n_iter=10, w=15, b=5, lr=0.5, sigma=3.0);
    create_neuron_map();
    t_bmu *cl = NULL;
    cl = train(p, 2, 128);

    printf("Over.\n");

	return 0;
}
*/

File: KTH_DS02/somdtw.py
import numpy as np
import matplotlib.pylab as plt
from matplotlib.gridspec import GridSpec
from sklearn.metrics import adjusted_rand_score
from numpy import inf
import math
import time
from scipy.io import arff
from scipy.special import comb
from collections import defaultdict

def DTWdist(x, y, w):
    # x (m*K) array; y (n*K) array
    # int w: window size limiting the maximal distance
    DTW ={}
    m = x.shape[0]
    n = y.shape[0]

    w = max(w, abs(m-n))
    for i in range(-1,m):
        for j in range(-1,n):
            DTW[(i, j)] = inf
    DTW[(-1, -1)] = 0
    for i in range(m):
        for j in range(max(0, i-w), min(n, i+w)):
            dist = (x[i]-y[j])**2
            DTW[(i, j)] = dist + min(DTW[(i-1, j)], DTW[(i, j-1)], DTW[(i-1, j-1)])
    return math.sqrt(DTW[m-1, n-1])


def LB_Keogh(x,y,b):
    # x (m*K) array; y (n*K) array
    # b is the window size
    LB_sum = 0
    for ind, i in enumerate(x):
        lower_bound = min(y[(ind-b if ind-b>=0 else 0):(ind+b)])
        upper_bound = max(y[(ind-b if ind-b>=0 else 0):(ind+b)])
        if i > upper_bound:
            LB_sum = LB_sum + (i-upper_bound)**2
        elif i < lower_bound:
            LB_sum = LB_sum + (i-lower_bound)**2

    return math.sqrt(LB_sum)


def dataloader(filename):
    data = arff.loadarff(filename)[0]
    data1 = np.array(list(data[0]))
    for i in data[1:]:
        data1 = np.vstack((data1,np.array(list(i))))

    return data1.astype('float')


def lr_decay(lr, t, n_iteration):
    return lr / (1 + t/(n_iteration/2))


class SOMdtw(object):
    def __init__(self, width, height, channel, sigma = 1.0, learning_rate=0.5, w=100, b=5):
        # Set up the SOM map (w,h,ch) and lr
        self.x = width
        self.y = height
        self.ch = channel
        self.lr = learning_rate
        self.sigma = sigma
        self.w = w
        self.b = b
        self.trained = False

    def initialize_map(self):
        # Initialize map weight vectors Kohonen random initialization
        ds_mul = np.mean(self.data) / 0.5
        self.node_vectors = np.random.rand(self.y, self.x, self.ch) * ds_mul

    def save(self, filename):
        # Save SOM to .npy file.
        if self.trained:
            np.savez(filename, x = self.node_vectors, w = np.array(self.w), b = np.array(self.b))
            return True
        else:
            return False

    def load(self, filename):
        # Load SOM from .npy file.
        npzfile = np.load(filename)
        self.node_vectors = npzfile['x']
        self.w = npzfile['w'][0]
        self.b = npzfile['b'][0]
        self.x = self.node_vectors.shape[1]
        self.y = self.node_vectors.shape[0]
        self.ch = self.node_vectors.shape[2]
        self.trained = True
        return True

    def get_map_vectors(self):
        # Returns the map vectors.
        if self.trained:
            return self.node_vectors
        else:
            return False

    def euc_distance(self, x, y):
        dist = np.linalg.norm(x - y)
        return dist

    def find_maching_nodes(self, input_sample):
        # This is to be called only when the map is trained.
        if self.trained == False:
            return False
        inf_start = time.time()
        n_data = input_sample.shape[0]
        locations = np.zeros((n_data, 3), dtype=np.int32)
        distances = np.zeros((n_data), dtype=np.float32)

        print_step = int(n_data / 20)
        print_count = 0
        for idx in range(n_data):

            if idx % print_step == 0:
                print_count += 1
                print('Finding mathing nodes:' + str(print_count))

            data_vect = input_sample[idx]
            min_dist = inf
            x = None
            y = None
            if self.dm == 1:
                for y_idx in range(self.y):
                    for x_idx in range(self.x):
                        node_vect = self.node_vectors[y_idx, x_idx]
                        if LB_Keogh(data_vect, node_vect, self.b) < min_dist:
                            dist = DTWdist(data_vect, node_vect, self.w)
                            if dist < min_dist:
                                min_dist = dist
                                x = x_idx
                                y = y_idx
            elif self.dm == 2:
                for y_idx in range(self.y):
                    for x_idx in range(self.x):
                        node_vect = self.node_vectors[y_idx, x_idx]
                        dist = self.euc_distance(data_vect, node_vect)
                        if dist < min_dist:
                            min_dist = dist
                            x = x_idx
                            y = y_idx

            locations[idx, 0] = y
            locations[idx, 1] = x
            locations[idx, 2] = x + self.x * y
            distances[idx] = min_dist

        print('Done')
        inf_end = time.time()
        print('Infering done in {:.6f} seconds.'.format(inf_end - inf_start))

        return locations, distances

    def train(self, data, n_iteration=10, batch_size=10, dm = 1):
    # Train the som, neighborhood function = gaussian, distance method = dtw
        self.data = data
        self.n_iter = n_iteration
        self.batch_size =batch_size
        self.dm = dm

        start_time = time.time()
        self.initialize_map()

        self.nb_dist = math.floor(min(self.x, self.y) / 1.5)
        n_samples = self.data.shape[0]
        data_index = np.arange(n_samples)

        # Pad the vector map to allow easy array processing.
        tmp_node_vects = np.zeros((self.y + 2 * self.nb_dist, self.x + 2 * self.nb_dist, self.ch))
        tmp_node_vects[self.nb_dist: self.nb_dist + self.y,
        self.nb_dist: self.nb_dist + self.x] = self.node_vectors.copy()
        self.node_vectors = tmp_node_vects

        batch_count = math.ceil(n_samples / self.batch_size)

        #self.neighbor_function()
        for iteration in range(self.n_iter):
            self.neighbor_function()
            np.random.shuffle(data_index)
            total_dist = 0

            # batch processing
            for batch in range(batch_count):
                steps_left = n_samples - batch * self.batch_size
                if steps_left < self.batch_size:
                    steps_in_batch = steps_left
                else:
                    steps_in_batch = self.batch_size
                # each sample is assigned with the best match node
                best_node = np.zeros((steps_in_batch, 3), dtype=np.int32)

                for step in range(steps_in_batch):
                    input_idx = data_index[batch * self.batch_size + step]
                    input_vect = self.data[input_idx]
                    y, x, dist = self.find_best_matching_node(input_vect)
                    best_node[step, 0] = y
                    best_node[step, 1] = x
                    best_node[step, 2] = input_idx
                    total_dist += dist
                # update the weights for each batch
                self.update_node_vectors(best_node)

            self.sigma = lr_decay(self.sigma, iteration+1, self.n_iter)
            self.lr = lr_decay(self.lr, iteration+1, self.n_iter)
            print('Iteration = {:d}, Average distance = {:.2f}.'.format(iteration+1,total_dist/n_samples))

        self.node_vectors = self.node_vectors[self.nb_dist : self.nb_dist + self.y, self.nb_dist : self.nb_dist + self.x]

        end_time = time.time()
        self.trained = True
        print('Training done in {:.6f} seconds.'.format(end_time - start_time))


    def find_best_matching_node(self, data_vect):
        # This method is used to find best matching node for data vector.
        # The node coordinates and distance are returned.
        min_dist = inf
        x = 0
        y = 0
        for y_idx in range(self.y):
            for x_idx in range(self.x):
                node_vect = self.node_vectors[y_idx + self.nb_dist, x_idx + self.nb_dist]
                if self.dm == 1:
                    if LB_Keogh(data_vect, node_vect, self.b) < min_dist:
                        dist = DTWdist(data_vect, node_vect, self.w)
                        if dist < min_dist:
                            min_dist = dist
                            x = x_idx
                            y = y_idx
                elif self.dm == 2:
                    dist = self.euc_distance(data_vect, node_vect)
                    if dist < min_dist:
                        min_dist = dist
                        x = x_idx
                        y = y_idx
        return y, x, min_dist

    def update_node_vectors(self, best_node):
        # This method updates the map node weights.
        for ind in range(best_node.shape[0]):
            node_y = best_node[ind, 0]
            node_x = best_node[ind, 1]
            input_ind = best_node[ind, 2]
            input_vec = self.data[input_ind]
            old_vec = self.node_vectors[node_y + self.y_delta + self.nb_dist, node_x + self.x_delta + self.nb_dist]

            update_vec = self.neighbor_weights * self.lr * (np.expand_dims(input_vec, axis=0) - old_vec)

            self.node_vectors[node_y + self.y_delta + self.nb_dist,
                              node_x + self.x_delta + self.nb_dist, :] += update_vec

    def neighbor_function(self):
        # Create a Guassian distribution matrix with (x,y,channel) and sigma is decayed according to iteration number
        # modify self.node_weights
        size = self.nb_dist * 2
        if size == 0:
            size = size + 2
        self.neighbor_weights = np.full((size * size, self.ch), 0.0)
        cp = size / 2.0
        p1 = 1.0 / (2 * math.pi * self.sigma ** 2)
        pdiv = 2.0 * self.sigma ** 2

        y_delta = []
        x_delta = []
        for y in range(size):
            for x in range(size):
                ep = -1.0 * ((x - cp) ** 2.0 + (y - cp) ** 2.0) / pdiv
                value = p1 * math.e ** ep
                self.neighbor_weights[y * size + x] = value
                y_delta.append(y - int(cp))
                x_delta.append(x - int(cp))
        self.x_delta = np.array(x_delta, dtype=np.int32)
        self.y_delta = np.array(y_delta, dtype=np.int32)
        #print(self.neighbor_weights, self.nb_dist)
        self.neighbor_weights -= self.neighbor_weights[size // 2]
        self.neighbor_weights[self.neighbor_weights < 0] = 0
        self.neighbor_weights /= np.max(self.neighbor_weights)


if __name__ == '__main__':
    w = 15
    b = 5
    name = 'BME'
    train = dataloader('./Univariate_arff/' + name + '/' + name + '_TRAIN.arff')
    test = dataloader('./Univariate_arff/' + name + '/' + name + '_TEST.arff')
    data = np.vstack((train[:,:-1], test[:,:-1]))
    label = np.hstack((train[:, -1], test[:, -1]))
    ch = data.shape[1]

    som1 = SOMdtw(width=3, height=1, channel=ch, sigma = 1.0, learning_rate=0.5, w=w, b=b)
    # dm = 1 : dtw; dm = 2 : Euc
    som1.train(data=data, n_iteration=10, batch_size=10, dm=2)
    samples = data.shape[0]

    # inference
    locations, distances = som1.find_maching_nodes(data)

    count = 0
    for i in range(samples):
        for j in range(i, samples):
            if i==j:
                count = count - 1
            if (label[i] == label[j]) and (locations[i,2] == locations[j,2]):
                count = count + 1
            elif (label[i] != label[j]) and (locations[i,2] != locations[j,2]):
                count = count + 1

    cn = comb(samples, 2)
    ri = count / cn

    print(cn, count, ri)
    print(adjusted_rand_score(label, locations[:, 2]))

    winmap = defaultdict(list)
    for i in range(samples):
        winmap[locations[i, 2]].append(data[i, :])

    # plt.rcParams['savefig.dpi'] = 500
    # plt.rcParams['figure.dpi'] = 500
    # fig = plt.figure(tight_layout=True)
    # # grid row : x, coloum : y
    # size1 = 3
    # size2 = 4
    # gs = GridSpec(size1, size2)
    # for x in range(size1):
        # for y in range(size2):
            # ax = fig.add_subplot(gs[x, y])
            # ki = y * size1 + x
            # if ki < len(winmap) and winmap[ki]:
                # ax.plot(np.mean(winmap[ki], axis=0))

            # ax.set_title('Cluster %d' % (ki + 1))

    # #plt.savefig('./scri6.png')
    # plt.show()